### PR TITLE
core(metrics): support CLS for all frames

### PR DESCRIPTION
**Pull request by adamraine**
Part 1a (step 2?) of the [Framehole implementation plan](https://docs.google.com/document/d/12bjRVM-kEMs6ChKyFQzCrVjdJiKKn65xZOcQvw643ME/edit?ts=5fb6f439#heading=h.7tf1bie89pvz) for CLS.

The `cumulative_score` arg in `LayoutShift` trace events is computed per-frame. We could find the latest `LayoutShift` event for each frame but it seemed simpler to compute our own cumulative score from every layout shift event.

Leaving this as a draft for now, will change the merge into branch to `master` once https://github.com/GoogleChrome/lighthouse/pull/11701 lands.

### CODE REVIEW

File: lighthouse-core/computed/metrics/cumulative-layout-shift-all-frames.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+const makeComputedArtifact = require('../computed-artifact.js');
+
+class CumulativeLayoutShiftAllFrames {
+  /**
+   * @param {LH.Trace} trace
+   * @return {Promise<{value: number}>}
+   */
+  static async compute_(trace) {
+    const cumulativeShift = trace.traceEvents
+      .filter(e => e.name === 'LayoutShift' && e.args && e.args.data && e.args.data.score)
```
Comment (connorjclark) [inline lighthouse-core/computed/metrics/cumulative-layout-shift-all-frames.js:17]: You can narrow the type by annotating the filter function to be a type guard. here's an example: https://github.com/GoogleChrome/lighthouse/pull/7817/files#diff-b66c6ac9b4b8c0726c736eb7479a9e4ff22c7b81bfa1b2b259c2cdfa220ea1f8R19

Comment (connorjclark) [reply lighthouse-core/computed/metrics/cumulative-layout-shift-all-frames.js:17]: oh nvm the type here is kinda complicated, a ts ignore on L19 is probably better

File: lighthouse-core/computed/metrics/timing-summary.js [hunk 2]
```diff
@@ -57,6 +59,9 @@ class TimingSummary {
     const cumulativeLayoutShiftValue = cumulativeLayoutShift &&
       cumulativeLayoutShift.value !== null ?
       cumulativeLayoutShift.value : undefined;
+    const cumulativeLayoutShiftAllFramesValue = cumulativeLayoutShiftAllFrames &&
+      cumulativeLayoutShiftAllFrames.value !== null ?
```
Comment (patrickhulce) [inline lighthouse-core/computed/metrics/timing-summary.js:63]: can your version ever be null? I don't *think* original CLS could be, but I'm even more certain with the reduce here that this one can't :)

File: lighthouse-core/lib/minify-trace.js [hunk 3]
```diff
@@ -48,6 +48,7 @@ const traceEventsToAlwaysKeep = new Set([
   // Needed for CPU profiler task attribution
   'Profile',
   'ProfileChunk',
+  'LayoutShift',
```
Comment (patrickhulce) [inline lighthouse-core/lib/minify-trace.js:51]: 👍 good find

File: lighthouse-core/test/audits/metrics-test.js [hunk 4]
```diff
@@ -99,6 +102,30 @@ describe('Performance: metrics', () => {
...
+        [MetricsAudit.DEFAULT_PASS]: clsAllFramesTrace,
+      },
+      devtoolsLogs: {
+        [MetricsAudit.DEFAULT_PASS]: clsAllFramesDevtoolsLog,
+      },
+    };
+
+    const context = {settings: {throttlingMethod: 'simulate'}, computedCache: new Map()};
+    const {details} = await MetricsAudit.audit(artifacts, context);
+    expect(details.items[0].cumulativeLayoutShift).toMatchInlineSnapshot(`0.0011656245471340055`);
```
Comment (patrickhulce) [inline lighthouse-core/test/audits/metrics-test.js:117]: I wouldn't expect these to ever change, `toBeCloseTo` instead?

File: lighthouse-core/test/fixtures/traces/frame-metrics-m89.devtools.log.json [hunk 5]
```diff
@@ -0,0 +1,53 @@
+[
```
Comment (patrickhulce) [inline lighthouse-core/test/fixtures/traces/frame-metrics-m89.devtools.log.json:1]: same trace as before? :)

Comment (adamraine) [reply lighthouse-core/test/fixtures/traces/frame-metrics-m89.devtools.log.json:1]: That one didn't have any layout shifts. Maybe replace the previous trace with this and use for both?

File: lighthouse-core/computed/metrics/cumulative-layout-shift-all-frames.js (hunk 1 above)
Comment (connorjclark) [reply lighthouse-core/computed/metrics/cumulative-layout-shift-all-frames.js:17]: totes forgot about the recent input thing. that was a rough bug

File: lighthouse-core/test/fixtures/traces/frame-metrics-m89.devtools.log.json (hunk 5 above)
Comment (patrickhulce) [reply lighthouse-core/test/fixtures/traces/frame-metrics-m89.devtools.log.json:1]: you read my mind :D

Comment (adamraine) [reply lighthouse-core/test/fixtures/traces/frame-metrics-m89.devtools.log.json:1]: I'm thinking merge https://github.com/GoogleChrome/lighthouse/pull/11701 and make the change here, sound good?

**Review by patrickhulce — APPROVED** [review]

**Comment by adamraine** [comment]
Both this and #11701 look ready to go. Any objections to merging this into #11701 and landing together?

**Comment by patrickhulce** [comment]
> Any objections to merging this into #11701 and landing together?

image

**Pull request merged**